cmd/collect_signals/main.go

```diff
@@ -0,0 +1,191 @@
+	"github.com/ossf/criticality_score/cmd/collect_signals/githubmentions"
+	"github.com/ossf/criticality_score/cmd/collect_signals/result"
+	"github.com/ossf/criticality_score/internal/githubapi"
+	"github.com/ossf/criticality_score/internal/logflag"
+	"github.com/ossf/criticality_score/internal/outfile"
+	"github.com/ossf/scorecard/v4/clients/githubrepo/roundtripper"
+	sclog "github.com/ossf/scorecard/v4/log"
+	log "github.com/sirupsen/logrus"
+	//csvutil (it's docs are heaps better!)
+)
+
+const defaultLogLevel = log.InfoLevel
+
+var logFlag = logflag.Level(defaultLogLevel)
+
+func init() {
+	flag.Var(&logFlag, "log", "set the `level` of logging.")
+	outfile.DefineFlags(flag.CommandLine, "force", "append", "OUT_FILE")
+	flag.Usage = func() {
+		cmdName := path.Base(os.Args[0])
+		w := flag.CommandLine.Output()
+		fmt.Fprintf(w, "Usage:\n  %s [FLAGS]... IN_FILE... OUT_FILE\n\n", cmdName)
+		fmt.Fprintf(w, "Collects signals for each project repository listed.\n")
+		fmt.Fprintf(w, "IN_FILE must be either a file or - to read from stdin.\n")
+		fmt.Fprintf(w, "OUT_FILE must be either be a file or - to write to stdout.\n")
+		fmt.Fprintf(w, "\nFlags:\n")
+		flag.PrintDefaults()
+	}
+}
+
+func main() {
+	flag.Parse()
+
+	logger := log.New()
+	logger.SetLevel(logFlag.Level())
+
+	// roundtripper requires us to use the scorecard logger.
+	scLogger := sclog.NewLogrusLogger(logger)
+
+	if flag.NArg() < 2 {
+		logger.Error("Must have at least one input file and an output file specified.")
+		os.Exit(2)
+	}
+	lastArg := flag.NArg() - 1
+
+	// Open all the in-files for reading
+	readers := []io.Reader{}
+	consumingStdin := false
+	for _, inFilename := range flag.Args()[:lastArg] {
+		if inFilename == "-" && !consumingStdin {
+			logger.Info("Reading from stdin")
+			// Only add stdin once.
+			consumingStdin = true
+			readers = append(readers, os.Stdin)
+			continue
+		}
+		logger.WithFields(log.Fields{
+			"filename": inFilename,
+		}).Debug("Reading from file")
+		f, err := os.Open(inFilename)
+		if err != nil {
+			logger.WithFields(log.Fields{
+				"error":    err,
+				"filename": inFilename,
+			}).Error("Failed to open an input file")
+			os.Exit(2)
+		}
+		defer f.Close()
+		readers = append(readers, f)
+	}
+	r := io.MultiReader(readers...)
+
+	// Open the out-file for writing
+	outFilename := flag.Args()[lastArg]
+	w, err := outfile.Open(outFilename)
+	if err != nil {
+		logger.WithFields(log.Fields{
+			"error":    err,
+			"filename": outFilename,
+		}).Error("Failed to open file for output")
+		os.Exit(2)
+	}
+	defer w.Close()
+
+	ctx := context.Background()
+
+	// Prepare a client for communicating with GitHub's GraphQLv4 API and Restv3 API
+	rt := roundtripper.NewTransport(ctx, scLogger)
+	httpClient := &http.Client{
+		Transport: rt,
+	}
+	ghClient := githubapi.NewClient(httpClient)
+
+	ghRepoFactory := github.NewRepoFactory(ghClient, logger)
+
+	// Register all the collectors that are supported.
+	collector.Register(&github.RepoCollector{})
```

Comment:
Would it make sense to do these register calls in the init() of the relevant packages instead?

Reply:
I considered that.

In some cases a collector will need parameters passed in - and I think it will be easier to see a full set of collectors. Additionally the order of calls to `Register()` is preserved in the output, which can't be guaranteed with `init()` so much.

Reply:
ack.

Reply:
FWIW I've added some more docs around the ordering.

`Collect()` will get some too soon, after I've re-implemented it to use goroutines.